tests/helpers/test_entity_registry.py

```diff
@@ -1451,6 +1582,204 @@ async def test_remove_config_entry_from_device_removes_entities_2(
+        remove_config_subentry_id=None,
+    )
+    await hass.async_block_till_done()
+
+    assert device_registry.async_get(device_entry.id)
+    assert not entity_registry.async_is_registered(entry_1.entity_id)
+    assert entity_registry.async_is_registered(entry_2.entity_id)
+    assert not entity_registry.async_is_registered(entry_3.entity_id)
+
+    # Remove the third config subentry from the device, the entity associated with it
+    # (and the device itself) should be removed
+    device_registry.async_update_device(
+        device_entry.id,
+        remove_config_entry_id=config_entry_1.entry_id,
+        remove_config_subentry_id="mock-subentry-id-2",
+    )
+    await hass.async_block_till_done()
+
+    assert not device_registry.async_get(device_entry.id)
+    assert not entity_registry.async_is_registered(entry_1.entity_id)
+    assert not entity_registry.async_is_registered(entry_2.entity_id)
+    assert not entity_registry.async_is_registered(entry_3.entity_id)
+
+
+async def test_remove_config_subentry_from_device_removes_entities_2(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    entity_registry: er.EntityRegistry,
+) -> None:
+    """Test that we don't remove entities with no config entry when device is modified."""
+    config_entry_1 = MockConfigEntry(
+        domain="hue",
+        subentries_data=[
+            config_entries.ConfigSubentryData(
+                data={},
+                subentry_id="mock-subentry-id-1",
+                title="Mock title",
+                unique_id="test",
+            ),
+            config_entries.ConfigSubentryData(
+                data={},
+                subentry_id="mock-subentry-id-2",
+                title="Mock title",
+                unique_id="test",
+            ),
+        ],
+    )
+    config_entry_1.add_to_hass(hass)
+
+    # Create device with three config subentries
+    device_registry.async_get_or_create(
+        config_entry_id=config_entry_1.entry_id,
+        config_subentry_id="mock-subentry-id-1",
+        connections={(dr.CONNECTION_NETWORK_MAC, "12:34:56:AB:CD:EF")},
+    )
+    device_registry.async_get_or_create(
+        config_entry_id=config_entry_1.entry_id,
+        config_subentry_id="mock-subentry-id-2",
+        connections={(dr.CONNECTION_NETWORK_MAC, "12:34:56:AB:CD:EF")},
+    )
+    device_entry = device_registry.async_get_or_create(
+        config_entry_id=config_entry_1.entry_id,
+        connections={(dr.CONNECTION_NETWORK_MAC, "12:34:56:AB:CD:EF")},
+    )
+    assert device_entry.config_entries == {config_entry_1.entry_id}
+    assert device_entry.config_entries_subentries == {
+        config_entry_1.entry_id: {None, "mock-subentry-id-1", "mock-subentry-id-2"},
+    }
+
+    # Create an entity without config entry or subentry
+    entry_1 = entity_registry.async_get_or_create(
+        "light",
+        "hue",
+        "5678",
+        device_id=device_entry.id,
```

Comment:
When does an integration connect its entities to another integration's device without having a config entry of its own?

I was thinking maybe the template integration but that only links to devices when set up from a config entry.

The entity platform helper code doesn't support connecting a device via `device_info` if the added entity doesn't have a config entry.

The only place in our codebase I found that created an entity entry directly via `async_get_or_create` without a config entry was the `TagEntity` which doesn't connect to a device entry.